Fix OOM

wlwr opened the pull request:
**What this PR does / why we need it**:

Fix memory leaks in high concurrent requests. 


In the case of higher QPS, the ingress-nginx-controller has serious memory leaks. After troubleshooting, the reason is that the stream_latency in `socket.go` uses the prometheus summary data type, the summary data type uses the global lock [#L277](https://github.com/kubernetes/ingress-nginx/blob/master/vendor/github.com/prometheus/client_golang/prometheus/summary.go#L277). the request volume is large, The global lock will cause a lot of gorontinue to wait, the memory cannot be released, and each gorontinue The variable allocation memory is large (variables contain up to 1w request information).

After upstream_latency is changed to use the prometheus Histogram data type, the CPU is reduced and there is no memory leak, because the Histogram logic is simple and no locks are used.

**Which issue this PR fixes** : 

fixes #3314

wlwr commented:
I signed it

wlwr commented:
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please email the CNCF helpdesk: [helpdesk@rt.linuxfoundation.org](mailto:helpdesk@rt.linuxfoundation.org)

I had applied for CLA

wlwr commented:
@aledbf 

In the case of 1W QPS, Analysis of goroutine by pprof tool.

```
File: nginx-ingress-controller
Type: goroutine
Time: Nov 17, 2018 at 11:09pm (CST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 1114, 99.73% of 1117 total
Dropped 76 nodes (cum <= 5)
Showing top 20 nodes out of 57
      flat  flat%   sum%        cum   cum%
      1107 99.10% 99.10%       1107 99.10%  runtime.gopark
         7  0.63% 99.73%          7  0.63%  syscall.Syscall6
         0     0% 99.73%          5  0.45%  internal/poll.(*pollDesc).wait
         0     0% 99.73%          5  0.45%  internal/poll.(*pollDesc).waitRead
         0     0% 99.73%          5  0.45%  internal/poll.runtime_pollWait
         0     0% 99.73%          6  0.54%  io.ReadAtLeast
         0     0% 99.73%       1035 92.66%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage
         0     0% 99.73%       1035 92.66%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage-fm
         0     0% 99.73%       1035 92.66%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.handleMessages
         0     0% 99.73%          6  0.54%  k8s.io/ingress-nginx/internal/watch.(*OSFileWatcher).watch.func1
         0     0% 99.73%       1035 92.66%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*summary).Observe
```

```
(pprof) peek Observe 
----------------------------------------------------------+-------------
      flat  flat%   sum%        cum   cum%   calls calls% + context
----------------------------------------------------------+-------------
                                              1035   100% |   k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage
         0     0%     0%       1035 92.66%                | k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*summary).Observe
                                              1034 99.90% |   sync.(*Mutex).Lock
                                                 1 0.097% |   k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*summary).asyncFlush
----------------------------------------------------------+-------------
```

Analysis of heap by pprof tool.

```
 1959.94MB 61.23% 61.23%  1959.94MB 61.23%  reflect.unsafe_NewArray
 1103.53MB 34.47% 95.70%  1224.03MB 38.24%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).ReadString
  120.50MB  3.76% 99.47%   120.50MB  3.76%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).readEscapedChar
         0     0% 99.47%  3184.97MB 99.50%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage
         0     0% 99.47%  3184.97MB 99.50%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage-fm
         0     0% 99.47%  3187.96MB 99.59%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.handleMessages
         0     0% 99.47%  3183.97MB 99.47%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).ReadVal
         0     0% 99.47%   120.50MB  3.76%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).readStringSlowPath
         0     0% 99.47%  3183.97MB 99.47%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*frozenConfig).Unmarshal
         0     0% 99.47%  1224.03MB 38.24%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*generalStructDecoder).Decode
         0     0% 99.47%  1224.03MB 38.24%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*generalStructDecoder).decodeOneField
         0     0% 99.47%   999.03MB 31.21%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*placeholderDecoder).Decode
         0     0% 99.47%  3183.97MB 99.47%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*sliceDecoder).Decode
         0     0% 99.47%  3183.97MB 99.47%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*sliceDecoder).doDecode
         0     0% 99.47%  1110.53MB 34.69%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*stringCodec).Decode
         0     0% 99.47%  1110.53MB 34.69%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*structFieldDecoder).Decode
         0     0% 99.47%  1959.94MB 61.23%  k8s.io/ingress-nginx/vendor/github.com/modern-go/reflect2.(*UnsafeSliceType).UnsafeGrow
         0     0% 99.47%  1959.94MB 61.23%  k8s.io/ingress-nginx/vendor/github.com/modern-go/reflect2.(*UnsafeSliceType).UnsafeMakeSlice
```

`summary.Observe` lock is serious.

aledbf commented:
@wlwr please post the difference of the metrics endpoint (to see exactly the change in the output)

wlwr commented:
@aledbf 

upstream_latency metrics logs


Before change

```
# HELP nginx_ingress_controller_ingress_upstream_latency_seconds Upstream service latency per Ingress
# TYPE nginx_ingress_controller_ingress_upstream_latency_seconds summary
nginx_ingress_controller_ingress_upstream_latency_seconds{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-g2zt7",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",quantile="0.5"} NaN
nginx_ingress_controller_ingress_upstream_latency_seconds{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-g2zt7",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",quantile="0.9"} NaN
nginx_ingress_controller_ingress_upstream_latency_seconds{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-g2zt7",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",quantile="0.99"} NaN
nginx_ingress_controller_ingress_upstream_latency_seconds_sum{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-g2zt7",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench"} 3.270999999999751
nginx_ingress_controller_ingress_upstream_latency_seconds_count{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-g2zt7",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench"} 1.221225e+06
```


After the change

```
# HELP nginx_ingress_controller_ingress_upstream_latency_seconds Upstream service latency per Ingress
# TYPE nginx_ingress_controller_ingress_upstream_latency_seconds histogram
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.005"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.01"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.025"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.05"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.1"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.25"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="0.5"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="1"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="2.5"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="5"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="10"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_bucket{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench",le="+Inf"} 35542
nginx_ingress_controller_ingress_upstream_latency_seconds_sum{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench"} 0.16500000000000012
nginx_ingress_controller_ingress_upstream_latency_seconds_count{controller_class="nginx",controller_namespace="ingress-nginx",controller_pod="nginx-ingress-controller-zdjzm",ingress="bifrost-bid-ingress-bench",namespace="bidex-comb",service="bifrost-bid-ingress-bench"} 35542
```

wlwr commented:
@aledbf 

Benchmark: Histogram VS Summary

Benchmark test code from the Prometheus

- https://github.com/prometheus/client_golang/blob/v0.8.0/prometheus/histogram_test.go
- https://github.com/prometheus/client_golang/blob/v0.8.0/prometheus/summary_test.go

```
go test -bench=. -benchtime 1s

BenchmarkHistogramObserve1-4    50000000            27.0 ns/op
BenchmarkHistogramObserve2-4    20000000           115 ns/op
BenchmarkHistogramObserve4-4    10000000           284 ns/op
BenchmarkHistogramObserve8-4     3000000           483 ns/op
BenchmarkHistogramWrite1-4       1000000          2333 ns/op
BenchmarkHistogramWrite2-4       1000000          3812 ns/op
BenchmarkHistogramWrite4-4        300000          6688 ns/op
BenchmarkHistogramWrite8-4        200000         13362 ns/op
BenchmarkSummaryObserve1-4       2000000           909 ns/op
BenchmarkSummaryObserve2-4        500000          3821 ns/op
BenchmarkSummaryObserve4-4        300000          9437 ns/op
BenchmarkSummaryObserve8-4        100000         14930 ns/op
BenchmarkSummaryWrite1-4          200000         10529 ns/op
BenchmarkSummaryWrite2-4           50000         23920 ns/op
BenchmarkSummaryWrite4-4           30000         49426 ns/op
BenchmarkSummaryWrite8-4           20000         96795 ns/op
```

BenchmarkSummaryObserve8 compares BenchmarkHistogramObserve8 with a performance difference of nearly 100 times. If the amount of concurrency is larger, the difference is more obvious.

aledbf commented:
@wlwr thank you for this PR but we cannot merge it. The reason is the fix you are proposing shifts the issue to the metric prometheus because of the increment of metrics.

wlwr commented:
> @wlwr thank you for this PR but we cannot merge it. The reason is the fix you are proposing shifts the issue to the metric prometheus because of the increment of metrics.

Because I didn't find a better solution. In the case of high RPS, the summary data type consumes more machine resources. unless the summary type is deleted or the storage type is changed.

aledbf commented:
@wlwr can you run the same tests using the image `quay.io/kubernetes-ingress-controller/nginx-ingress-controller:dev` ?

aledbf commented:
> Because I didn't find a better solution. In the case of high RPS, the summary data type consumes more machine resources. unless the summary type is deleted or the storage type is changed.

I think #3439 will fix the lock issue you found in prometheus and also avoid the unbound creation of goroutines to process the stats from lua (the image in my previous comment contains the mentioned PR)

AlexPereverzyev commented:
@aledbf, I think, in high RPS scenarios it's more desirable to have better CPU/RAM utilization than detailed latency metrics. Whatever the fix will be, I hope it will not compromise Nginx performance.

aledbf commented:
> Whatever the fix will be, I hope it will not compromise Nginx performance.

Agree. We are aware of the current issues with the metrics in high RPS scenarios or thousand ingress rules/endpoints. In 0.22 we are going to allow refactor and improve the metrics

aledbf commented:
> Whatever the fix will be, I hope it will not compromise Nginx performance.

In #3439 we only change the way the ingress controller (go binary) processes the metrics

yvespp commented:
@aledbf a switch to turn off detailed metrics per ingress resource would be nice. We currently only use summs over all ingresses for monitoring. If we need more details we can still use the access logs.

wlwr commented:
> @wlwr can you run the same tests using the image `quay.io/kubernetes-ingress-controller/nginx-ingress-controller:dev` ?

I tried this image, there is still a memory leak and the CPU usage is high. 

It is recommended that the processor be pushed to the pool before json decoding to avoid pre-allocating array memory.

```
top - 13:16:38 up 5 days, 19:12,  1 user,  load average: 2.55, 2.70, 2.80
Tasks: 363 total,   1 running, 362 sleeping,   0 stopped,   0 zombie
%Cpu(s):  8.8 us,  2.3 sy,  0.0 ni, 87.9 id,  0.0 wa,  0.0 hi,  0.9 si,  0.0 st
KiB Mem : 24743123+total, 22361971+free, 13499860 used, 10311652 buff/cache
KiB Swap:        0 total,        0 free,        0 used. 23256392+avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
22446 33        20   0 9607440   9.1g  13176 S 110.6  3.9 143:11.90 nginx-ingress-c
23540 33        20   0  478420  83308   2372 S   9.2  0.0   7:19.38 nginx
 4486 root      20   0  213160  27044   3948 S   8.9  0.0 314:00.05 iotop
23546 33        20   0  478196  83076   2372 S   8.9  0.0   7:09.09 nginx
23908 33        20   0  478244  82916   2372 S   8.6  0.0   7:09.62 nginx
23535 33        20   0  478180  82960   2372 S   7.2  0.0   7:01.42 nginx
23537 33        20   0  477796  82676   2372 S   6.8  0.0   7:15.29 nginx
23548 33        20   0  477956  82720   2372 S   6.8  0.0   7:12.23 nginx
23543 33        20   0  478308  83204   2372 S   6.5  0.0   7:07.37 nginx
23547 33        20   0  478204  83192   2372 S   6.5  0.0   7:14.26 nginx
23902 33        20   0  478052  83132   2372 S   6.5  0.0   7:14.57 nginx
23613 33        20   0  478052  83088   2372 S   6.2  0.0   7:00.68 nginx
24063 33        20   0  478052  82892   2372 S   6.2  0.0   7:09.23 nginx
23545 33        20   0  478104  82960   2372 S   5.8  0.0   7:15.21 nginx
23688 33        20   0  478312  83032   2372 S   5.8  0.0   7:08.24 nginx
```


Analysis of goroutine by pprof tool.

```
File: nginx-ingress-controller
Type: goroutine
Time: Nov 21, 2018 at 1:17pm (CST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 41024, 100% of 41036 total
Dropped 124 nodes (cum <= 205)
      flat  flat%   sum%        cum   cum%
     41024   100%   100%      41024   100%  runtime.gopark
         0     0%   100%      40951 99.79%  k8s.io/ingress-nginx/vendor/gopkg.in/go-playground/pool%2ev3.(*limitedPool).Queue.func1
         0     0%   100%      40951 99.79%  runtime.chansend
         0     0%   100%      40951 99.79%  runtime.chansend1
         0     0%   100%      40989 99.89%  runtime.goparkunlock
```

Analysis of heap by pprof tool.

```
File: nginx-ingress-controller
Type: inuse_space
Time: Nov 21, 2018 at 1:18pm (CST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 6383.56MB, 99.33% of 6426.42MB total
Dropped 74 nodes (cum <= 32.13MB)
      flat  flat%   sum%        cum   cum%
 3897.73MB 60.65% 60.65%  3897.73MB 60.65%  reflect.unsafe_NewArray
 2271.56MB 35.35% 96.00%  2483.56MB 38.65%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).ReadString
     212MB  3.30% 99.30%      212MB  3.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).readEscapedChar
    2.26MB 0.035% 99.33%  6393.06MB 99.48%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage
         0     0% 99.33%  6393.06MB 99.48%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage-fm
         0     0% 99.33%  6397.83MB 99.55%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.handleMessages
         0     0% 99.33%  6381.29MB 99.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).ReadVal
         0     0% 99.33%      212MB  3.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).readStringSlowPath
         0     0% 99.33%  6381.29MB 99.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*frozenConfig).Unmarshal
         0     0% 99.33%  2483.56MB 38.65%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*generalStructDecoder).Decode
         0     0% 99.33%  2483.56MB 38.65%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*generalStructDecoder).decodeOneField
         0     0% 99.33%  2012.05MB 31.31%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*placeholderDecoder).Decode
         0     0% 99.33%  6381.29MB 99.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*sliceDecoder).Decode
         0     0% 99.33%  6381.29MB 99.30%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*sliceDecoder).doDecode
         0     0% 99.33%  2244.05MB 34.92%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*stringCodec).Decode
         0     0% 99.33%  2244.05MB 34.92%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*structFieldDecoder).Decode
         0     0% 99.33%  3897.73MB 60.65%  k8s.io/ingress-nginx/vendor/github.com/modern-go/reflect2.(*UnsafeSliceType).UnsafeGrow
         0     0% 99.33%  3897.73MB 60.65%  k8s.io/ingress-nginx/vendor/github.com/modern-go/reflect2.(*UnsafeSliceType).UnsafeMakeSlice
```

Analysis of cpu by pprof tool.

```
File: nginx-ingress-controller
Type: cpu
Time: Nov 21, 2018 at 1:18pm (CST)
Duration: 30.13s, Total samples = 33.33s (110.62%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top20
Showing nodes accounting for 31.85s, 95.56% of 33.33s total
Dropped 175 nodes (cum <= 0.17s)
Showing top 20 nodes out of 36
      flat  flat%   sum%        cum   cum%
    29.90s 89.71% 89.71%     29.90s 89.71%  runtime.memmove
     0.36s  1.08% 90.79%      0.36s  1.08%  k8s.io/ingress-nginx/vendor/github.com/beorn7/perks/quantile.NewTargeted.func1
     0.34s  1.02% 91.81%      0.34s  1.02%  unicode/utf8.ValidString
     0.31s  0.93% 92.74%      0.31s  0.93%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.hashAdd
     0.29s  0.87% 93.61%     30.47s 91.42%  k8s.io/ingress-nginx/vendor/github.com/beorn7/perks/quantile.(*stream).compress
     0.11s  0.33% 93.94%      0.24s  0.72%  runtime.mapaccess1_faststr
     0.10s   0.3% 94.24%     30.60s 91.81%  k8s.io/ingress-nginx/vendor/github.com/beorn7/perks/quantile.(*stream).merge
     0.09s  0.27% 94.51%      1.11s  3.33%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*metricVec).hashLabels
     0.09s  0.27% 94.78%      0.24s  0.72%  runtime.mallocgc
     0.07s  0.21% 94.99%      0.27s  0.81%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.matchLabels
     0.05s  0.15% 95.14%      0.23s  0.69%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*Iterator).ReadString
     0.05s  0.15% 95.29%      0.57s  1.71%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.validateValuesInLabels
     0.03s  0.09% 95.38%      1.90s  5.70%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).runUpdate.func1
     0.02s  0.06% 95.44%     30.64s 91.93%  k8s.io/ingress-nginx/vendor/github.com/beorn7/perks/quantile.(*Stream).insert
     0.01s  0.03% 95.47%     30.65s 91.96%  k8s.io/ingress-nginx/vendor/github.com/beorn7/perks/quantile.(*Stream).Insert
     0.01s  0.03% 95.50%      0.43s  1.29%  k8s.io/ingress-nginx/vendor/github.com/json-iterator/go.(*generalStructDecoder).decodeOneField
     0.01s  0.03% 95.53%      0.29s  0.87%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*metricMap).getMetricWithHashAndLabels
     0.01s  0.03% 95.56%      1.51s  4.53%  k8s.io/ingress-nginx/vendor/github.com/prometheus/client_golang/prometheus.(*metricVec).getMetricWith
         0     0% 95.56%      0.55s  1.65%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage
         0     0% 95.56%      0.55s  1.65%  k8s.io/ingress-nginx/internal/ingress/metric/collectors.(*SocketCollector).handleMessage-fm
```

iakisme commented:
Any update?

arthurh4 commented:
Are there any update on this ? we are also reaching this limitation.
